### PR TITLE
1488 Passable redux

### PR DESCRIPTION
**Pull request by turadg**
closes: #1488

## Description

Reattempting:
-  https://github.com/endojs/endo/pull/1933

Again:
- Makes `Passable` a generic type instead of `any`.
- Defines overloads for `passStyleOf` to return the actual style.
- Defines the `Key` in terms of `Passable`
- Makes a ton of fixes and suppressions in places that relied on the previous `any`'s

### Security Considerations

n/a

### Scaling Considerations

n/a

### Documentation Considerations

Better types are better documentation.

### Testing Considerations

- [ ] https://github.com/Agoric/agoric-sdk/pull/8774

### Upgrade Considerations

These changes may cause type errors downstream. I don't think we should call those breaking change à la semver because they don't affect the runtime.

### CODE REVIEW

**Comment by erights** [comment]
(I just reran the windows-latest flake)

**Comment by erights** [comment]
> These changes may cause type errors downstream. I don't think we should call those breaking change à la semver because they don't affect the runtime.

@kriskowal , I defer to you on this. I'll just say I hope we can adopt that precedent as a general rule ;)

File: packages/marshal/src/rankOrder.js
Comment (erights) [inline packages/marshal/src/rankOrder.js:145]: Just curious if you have a plan for how we can eventually get these dispatches to do the narrowing for us. Everything I thought of is worse that just leaving these `@ts-expect-error FIXME`s in place.

File: packages/marshal/src/rankOrder.js
Comment (erights) [inline packages/marshal/src/rankOrder.js:302]: What do you think of
```suggestion
 * @template {Passable} [T=Passable]
```
for all or most of these? It offloads the use site from needing to say `<Passable>` when that's all it would say anyway. I've been doing essentially this elsewhere and been happy with the results.

File: packages/pass-style/package.json
Comment (erights) [inline packages/pass-style/package.json:22]: Interesting. Do you expect this precedent to spread?

File: packages/pass-style/src/make-far.js
Comment (erights) [inline packages/pass-style/src/make-far.js:78]: Because this type expression is already do large, factoring out the `@import` seems even more compelling here.

File: packages/pass-style/src/passStyle-helpers.js
Comment (erights) [inline packages/pass-style/src/passStyle-helpers.js:33]: I've never seen this inline before. Nice.

File: packages/pass-style/src/passStyleOf.js
Comment (erights) [inline packages/pass-style/src/passStyleOf.js:110]:
```suggestion
        const innerStyle = passStyleMemo.get(inner);
        if (innerStyle) {
          return innerStyle;
        }
```

File: packages/pass-style/src/typeGuards.js
Comment (erights) [inline packages/pass-style/src/typeGuards.js:36]: Yeah, I don't remember why I have these separate named @callable types. Good riddance, thanks.

File: packages/patterns/src/keys/checkKey.js
Comment (erights) [inline packages/patterns/src/keys/checkKey.js:131]: Shouldn't this be

```suggestion
/**
 * @param {any} val
 * @returns {val is Key}
 */
```
?

File: packages/patterns/test/types.test-d.ts
Comment (erights) [inline packages/patterns/test/types.test-d.ts:5]: ;)

**Review by erights — APPROVED** [review]
LGTM, thanks!

**Comment by erights** [comment]
Did we stop requiring conventional commits?

File: packages/marshal/src/encodeToCapData.js
Comment (turadg) [reply packages/marshal/src/encodeToCapData.js:118]: it was `any` before because `Passable` was `any`. Making it `Passable` now causes an error at just about every access in this function. The job of this function is to inspect what the object actually is, narrow the type, and delegate. So the type hint isn't hindering more than helping.

File: packages/marshal/src/rankOrder.js
Comment (turadg) [reply packages/marshal/src/rankOrder.js:145]: There's something in TS for narrowing by `typeof` but I don't expect they'll our helper function for pass style. The pragmatic thing may be to drop to `any` in places where the code is doing its own type validation.

File: packages/marshal/src/rankOrder.js
Comment (turadg) [reply packages/marshal/src/rankOrder.js:302]: that makes sense for parameterized types, but here the template slot's type is constrained by the first argument. Giving a default doesn't help.

File: packages/pass-style/package.json
Comment (turadg) [reply packages/pass-style/package.json:22]: I'd rather export the types at the top level but I think that will require either manually maintaining an `index.d.ts` next to the `index.js` or having a single `index.ts` that can export both JS code and TS types. But the latter would entail a build step and thus require some new workflows and buy-in from the repo maintainers

File: packages/patterns/src/keys/checkKey.js
Comment (turadg) [reply packages/patterns/src/keys/checkKey.js:131]: indeed

File: packages/pass-style/package.json
Comment (mhofman) [reply packages/pass-style/package.json:22]: I'm confused, why can't `index.js` do `export * from './src/types.js'` ?

Comment (turadg) [reply packages/pass-style/package.json:22]: yeah, I realized that soon after my comment. it does require an eslint suppression but I'll do that

File: packages/marshal/src/encodeToCapData.js
Comment (erights) [reply packages/marshal/src/encodeToCapData.js:118]: Thx. Resolving.

File: packages/marshal/src/rankOrder.js
Comment (erights) [reply packages/marshal/src/rankOrder.js:145]: Thx. Resolving.

**Comment by erights** [comment]
> Did we stop requiring conventional commits?

My mistake. All your commits are fine. I was misled by the title, as my habit is to have the title also follow this convention (and usually be identical to the first commit).

File: packages/exo/index.js
Comment (kriskowal) [inline packages/exo/index.js:3]: That’s unfortunate and might explain some of my experience.

File: packages/exo/src/types.js
Comment (kriskowal) [inline packages/exo/src/types.js:1]: 👍

File: packages/far/test/test-marshal-far-obj.js
Comment (kriskowal) [inline packages/far/test/test-marshal-far-obj.js:9]: Would I be right to assume that someday `eslint` will catch up with this notation and demand spaces inside the curlies?

File: packages/marshal/src/deeplyFulfilled.js
Comment (kriskowal) [inline packages/marshal/src/deeplyFulfilled.js:14]: Optional wordsmithery.
```suggestion
// TODO return a type contingent on the parameter as deeplyFullfilledObject from agoric-sdk does
```

File: packages/marshal/src/encodePassable.js
Comment (kriskowal) [inline packages/marshal/src/encodePassable.js:731]:
```suggestion
  // @ts-expect-error Type 'unknown' is not assignable to type 'Passable<PassableCap, Error>'.
```
I think this is fine without elaboration.

File: packages/marshal/src/encodeToCapData.js
Comment (kriskowal) [reply packages/marshal/src/encodeToCapData.js:118]: Would `unknown` work better than `any` here, or is it better to blanket-ignore the type-unsafety internally?

File: packages/marshal/src/encodeToSmallcaps.js
Comment (kriskowal) [inline packages/marshal/src/encodeToSmallcaps.js:389]: What’s the fix? Is this as good as it gets?

File: packages/marshal/src/marshal.js
Comment (kriskowal) [inline packages/marshal/src/marshal.js:373]: 👆

File: packages/marshal/src/marshal.js
Comment (kriskowal) [inline packages/marshal/src/marshal.js:375]: 👆

File: packages/marshal/src/rankOrder.js
Comment (kriskowal) [inline packages/marshal/src/rankOrder.js:149]: 👆 Is the intention to address all FIXMEs before landing? These could be links to an issue, otherwise.

File: packages/marshal/src/rankOrder.js
Comment (kriskowal) [inline packages/marshal/src/rankOrder.js:156]: I would not object to ditto if these stick around.
```suggestion
          // @ts-expect-error ditto, type narrowed
```

File: packages/pass-style/src/passStyleOf.js
Comment (kriskowal) [inline packages/pass-style/src/passStyleOf.js:171]: 👆

File: packages/patterns/src/keys/compareKeys.js
Comment (kriskowal) [inline packages/patterns/src/keys/compareKeys.js:136]: 👆 I’m fine with a link and dittos if these are intended to commit. (but don’t like `FIXME` because I like `FIXME` and `XXX` to be used as markers for work that needs to be ironed out before a merge.

**Review by kriskowal — CHANGES_REQUESTED** [review]
I’m substantially excited about this change. I would like to break out a PR for the change to the release and build process so that it can suffer a revert separately if necessary. I’d also like to replace temporary markers `XXX` and `FIXME` with something but it doesn’t have to be tedious. It could be a single explanation or link to a tracking issue with a bunch of dittos or daggers for the repeated cases.

⚠️I will not die on this hill and I will approve this change regardless⚠️: My intuition (born from typed languages and sympathy for dependant projects that have to adapt when they upgrade this dependency) is that we should consider this a breaking change both in spirit and semver, and rush the necessary adaptations in Agoric SDK and dapps to minimize the pain. Since we are treading heavily over breaking changes with the “sync endo versions” workflow and `resolutions` farther out, there’s not much reason to spare the major version bump.

**Comment by kriskowal** [comment]
Also, please add `NEWS.md` with a description of what you’ve done for https://github.com/Agoric/agoric-sdk/pull/8774 so we can refer developers to this when they upgrade their *apps.

File: packages/base64/package.json
Comment (turadg) [reply packages/base64/package.json:35]: 👍 https://github.com/endojs/endo/pull/2258

File: packages/captp/src/captp.js
Comment (turadg) [reply packages/captp/src/captp.js:328]: bad comment. the type was wrong so I just removed it

File: packages/captp/src/captp.js
Comment (turadg) [reply packages/captp/src/captp.js:417]: solved

File: packages/exo/index.js
Comment (turadg) [reply packages/exo/index.js:3]: yeah. I think we should consider disabling the rule. I think it has more false positives and fewer true positives than TS

File: packages/marshal/src/encodeToSmallcaps.js
Comment (turadg) [reply packages/marshal/src/encodeToSmallcaps.js:389]: clarified

File: packages/marshal/src/rankOrder.js
Comment (turadg) [reply packages/marshal/src/rankOrder.js:149]: not before landing. maybe never. I've taken out the FIXME

**Comment by turadg** [comment]
I had to rebase to resolve merge conflicts created by https://github.com/endojs/endo/pull/2256

I squashed the fixups and rolled in other changes requested by review

**Review by kriskowal — APPROVED** [review]
You may want to look over the remaining XXX and TODO cases.

**Pull request merged**
